include/cantera/base/Solution.h

```diff
@@ -0,0 +1,74 @@
+//! @file Solution.h
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef CT_SOLUTION_H
+#define CT_SOLUTION_H
+
+#include "cantera/base/ctexceptions.h"
+
+namespace Cantera
+{
+
+class ThermoPhase;
+class Kinetics;
+class Transport;
+
+//! A container class holding managers for all pieces defining a phase
+class Solution : public std::enable_shared_from_this<Solution>
+{
+public:
+    Solution();
+    Solution(const std::string& infile, const std::string& phasename);
+    ~Solution() {}
+    Solution(const Solution&) = delete;
+    Solution& operator=(const Solution&) = delete;
+
+    static shared_ptr<Solution> create() {
+        return shared_ptr<Solution>( new Solution );
+    }
+
+    //! Return the name of this Solution object
+    std::string name() const;
+
+    //! Set the name of this Solution object
+    void setName(const std::string& name);
+
+    //! Generate self-documenting YAML string
+    std::string toYAML() const;
+
+    //! Set the ThermoPhase object
+    void setThermoPhase(shared_ptr<ThermoPhase> thermo);
+
+    //! Set the Kinetics object
+    void setKinetics(shared_ptr<Kinetics> kinetics);
+
+    //! Set the Transport object
+    void setTransport(shared_ptr<Transport> transport);
+
+    //! Accessor for the ThermoPhase object
+    shared_ptr<ThermoPhase> thermo() {
+        return m_thermo;
+    }
```

Comment:
Hm. Tried returning `m_thermo.get()`, but casting a smart pointer to `ThermoPhase&` appears to involve more work than expected. Anything I'm overlooking?

Reply:
> Hm. Tried this via `m_thermo.get()`, but casting a smart pointer to `ThermoPhase&` appears to involve more work than expected. Anything I'm overlooking? 

All you have to do to be able to return `ThermoPhase&` is dereference the shared_ptr, i.e.
```c++
ThermoPhase& thermo() {
    return *m_thermo;
}
```
 > (beyond your question being to revert to conventional pointers?)
No, I'm definitely not suggesting that `Solution` should store a raw pointer. Thinking about it more, the scenario of concern with returning a reference would where a user:
1. Allows the `ThermoPhase` object to be held only by the Solution object
2. Stores the address of that reference (i.e. the address of the underlying `ThermoPhase`)
3. Allows the `Solution` object to be destroyed
4. Tries to continue using the pointer to the `ThermoPhase` object.

I don't know how likely this is -- if a user takes the second step, I would argue that they are deliberately subverting the fact that the interface returns a reference, but perhaps it's better to avoid the possibility in any case.

> In the case of a 0D or 1D object, the `Solution` object could be used during initialization; however, it may be more efficient to store references to `ThermoPhase`, `Kinetics`, etc. internally. In either case, the base object can always be retrieved using the `m_root` pointer (see current `Phase::name` implementation).

Are you expecting that we will eventually require 0D / 1D objects to be initialized using `Solution` objects? Otherwise, what behavior should be exhibited when the `ThermoPhase` etc. used in a reactor network don't have an `m_root` pointer when attempting to serialize?

Reply:
:+1:  thanks for the code sample, I guess I was over-thinking this one! (I thought that smart pointers required something less straight-forward :man_facepalming:). Either way, I am changing the accessor as it isn't currently used (I simply added it due to being asked about it). Other accessors can be added later on as needed ...

> Are you expecting that we will eventually require 0D / 1D objects to be initialized using `Solution` objects? Otherwise, what behavior should be exhibited when the `ThermoPhase` etc. used in a reactor network don't have an m_root pointer when attempting to serialize?

Not sure where this may go in the long run: setting all in one swoop is simpler (regardless of the accessor implementation), and it's always possible to assign a root node with empty `Transport` and `Kinetics` managers (much like is already done in Python).